clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,63 +348,560 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+                                    "state into " + state;
+                        } else {
+                            reason = "The member has re-joined the group.";
+                        }
+                        // Revocation callback completed but the reconciled assignment should not
+                        // be applied (not relevant anymore). This could be because the member
+                        // is not in the RECONCILING state anymore (fenced, failed, unsubscribed),
+                        // or because it has already re-joined the group.
+                        CompletableFuture<Void> res = new CompletableFuture<>();
+                        res.completeExceptionally(new KafkaException("Interrupting reconciliation" +
+                                " after revocation. " + reason));
+                        return res;
+                    }
+                });
+
+        reconciliationResult.whenComplete((result, error) -> {
+            markReconciliationCompleted();
+            if (error != null) {
+                // Leaving member in RECONCILING state after callbacks fail. The member
+                // won't send the ack, and the expectation is that the broker will kick the
+                // member out of the group after the rebalance timeout expires, leading to a
+                // RECONCILING -> FENCED transition.
+                log.error("Reconciliation failed. ", error);
+            } else {
+                if (state == MemberState.RECONCILING) {
+
+                    // Make assignment effective on the broker by transitioning to send acknowledge.
+                    transitionTo(MemberState.ACKNOWLEDGING);
+
+                    // Make assignment effective on the member group manager
+                    currentAssignment = assignedPartitions;
+
+                    // Indicate that we completed reconciling a subset of the assignment ready to
+                    // reconcile (new assignments might have been received or discovered in
+                    // metadata)
+                    assignmentReadyToReconcile.removeAll(assignedPartitions);
+
+                } else {
+                    log.debug("New assignment processing completed but the member already " +
+                            "transitioned out of the reconciliation state into {}. Interrupting " +
+                            "reconciliation as it's not relevant anymore,", state);
+                    // TODO: double check if subscription state changes needed. This is expected to be
+                    //  the case where the member got fenced, failed or unsubscribed while the
+                    //  reconciliation was in process. Transitions to those states update the
+                    //  subscription state accordingly so it shouldn't be necessary to make any changes
+                    //  to the subscription state at this point.
```

Comment:
I agree that the other state transition should take care of updating the state. We should only abort here.

Reply:
Cool, thanks for confirming.